File: app/src/main/java/com/ericho/coupleshare/act/MainActivity3.java
package com.ericho.coupleshare.act;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.View;
import android.widget.Toast;

import com.ericho.coupleshare.R;
import com.ericho.coupleshare.adapter.HomePageAdapter;
import com.ericho.coupleshare.frag.BaseFrag;
import com.ericho.coupleshare.interf.FabListener;
import com.ericho.coupleshare.util.ThreadUtil;

import butterknife.BindView;
import butterknife.ButterKnife;
import timber.log.Timber;

public class MainActivity3 extends AppCompatActivity implements ViewPager.OnPageChangeListener {
    private static final int REQ_LOGIN = 111;


    @BindView(R.id.toolbar)
    protected Toolbar toolbar;
    @BindView(R.id.tabLayout)
    protected TabLayout tabLayout;
    @BindView(R.id.viewPager)
    protected ViewPager viewPager;
    @BindView(R.id.fab)
    protected FloatingActionButton floatingActionButton;

    private HomePageAdapter homePageAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        init();
        boolean isLogined = getUserLogined();
        if(!isLogined){
            Intent intent = new Intent(this,LoginAct.class);
            startActivityForResult(intent,REQ_LOGIN);
        }

    }

    private void init() {
        ButterKnife.bind(this);

        setSupportActionBar(toolbar);
        homePageAdapter = new HomePageAdapter(getSupportFragmentManager(),this);
        viewPager.setAdapter(homePageAdapter);
        tabLayout.setupWithViewPager(viewPager);
        viewPager.addOnPageChangeListener(this);
        this.onPageSelected(0);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.more,menu);
        return true;
    }

    private boolean getUserLogined() {
        return true;
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode){
            case REQ_LOGIN:
                if(resultCode == Activity.RESULT_CANCELED){
                    Toast.makeText(this,"Login canceled",Toast.LENGTH_SHORT).show();
                }else {
                    loadUserData();
                }
                break;
            default:
                super.onActivityResult(requestCode, resultCode, data);
        }
    }

    private void loadUserData() {
        //load user name and server photo's in background....
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

    }

    @Override
    public void onPageSelected(int position) {
        invalidateOptionsMenu();
        Timber.d("onPageSelected pos $d",position);
        Fragment fragment = homePageAdapter.getItem(position);
        if(fragment instanceof FabListener){
            //// TODO: 3/7/2017
            FabListener lis = (FabListener) fragment;
            floatingActionButton.setVisibility(View.VISIBLE);
            floatingActionButton.setOnClickListener(null);
            lis.onAttachFloatingActionListener(floatingActionButton);
        }
    }

    @Override
    public void onPageScrollStateChanged(int state) {

    }
}

File: app/src/main/java/com/ericho/coupleshare/act/MainActivity3.kt
package com.ericho.coupleshare.act

import android.app.Activity
import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.TabLayout
import android.support.v4.view.ViewPager
import android.support.v7.widget.Toolbar
import android.view.Menu
import android.view.MenuItem
import butterknife.bindView
import com.ericho.coupleshare.App
import com.ericho.coupleshare.Injection
import com.ericho.coupleshare.R
import com.ericho.coupleshare.adapter.HomePageAdapter
import com.ericho.coupleshare.constant.BroadcastConstant
import com.ericho.coupleshare.frag.LocationShowFrag2
import com.ericho.coupleshare.interf.PermissionListener
import com.ericho.coupleshare.mvp.data.LoginRepository
import com.ericho.coupleshare.mvp.presenter.LocationsPresenter
import com.ericho.coupleshare.service.LocationMonitorSer
import org.jetbrains.anko.toast
import timber.log.Timber
import java.util.Calendar


/**
 * Created by steve_000 on 8/7/2017.
 * for project CoupleShare
 * package name com.ericho.coupleshare.act
 */
class MainActivity3 : BasePermissionActivity(), ViewPager.OnPageChangeListener {


  val toolbar: Toolbar by bindView(R.id.toolbar)
  val tabLayout: TabLayout by bindView(R.id.tabLayout)
  val viewPager: ViewPager by bindView(R.id.viewPager)

  var loginRepository: LoginRepository? = null

  var mLocationPresenter: LocationsPresenter? = null
  var alarmManager: AlarmManager? = null
  var alarmLocationIntent: PendingIntent? = null
  private var homePageAdapter: HomePageAdapter? = null


  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)
    init()

    doNormalWork()

  }

  private fun doNormalWork() {
    this.checkSelfPermission(REQ_LOCATION_UPDATE, android.Manifest.permission.ACCESS_COARSE_LOCATION.toList(),
            listener = object : PermissionListener {
              override fun onGranted() {
                toast("granted ")
                startService(Intent(this@MainActivity3, LocationMonitorSer::class.java))
              }

              override fun onDenied(deniedPermission: List<String>) {
                val dialog = com.ericho.coupleshare.frag.AlertDialogFrag.newInstance("Error", "the following permission deny:\n" + "$deniedPermission")
                dialog.show(supportFragmentManager, "error")
              }
            })


  }

  private fun init() {

    setSupportActionBar(toolbar)
    homePageAdapter = HomePageAdapter(supportFragmentManager, this)
    viewPager.adapter = homePageAdapter
    tabLayout.setupWithViewPager(viewPager)
    viewPager.addOnPageChangeListener(this)
    this.onPageSelected(0)


    //for presneter
    mLocationPresenter = LocationsPresenter(
            homePageAdapter!!.getItem(2) as LocationShowFrag2,
            Injection.provideLocationsRepository(context = this))
    mLocationPresenter!!.start()

    loginRepository = Injection.provideLoginRepository(this.applicationContext)


    alarmManager = this.getSystemService(Context.ALARM_SERVICE) as AlarmManager
    val intent = Intent()
    intent.action = BroadcastConstant.REQ_LOC_MONITOR
    alarmLocationIntent = PendingIntent.getBroadcast(App.context, 0, intent, 0)

    alarmManager!!.setInexactRepeating(
            AlarmManager.ELAPSED_REALTIME,
            Calendar.getInstance().timeInMillis,
            10 * 60 * 1000,
            alarmLocationIntent)
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    menuInflater.inflate(R.menu.more, menu)
    menuInflater.inflate(R.menu.logout, menu)
    return true
  }

  override fun onOptionsItemSelected(item: MenuItem?): Boolean {
    when (item?.itemId) {
      R.id.more -> openSettingAct()
      R.id.menu_logout -> {
        loginRepository?.logout(this)
        startActivity(Intent(this, LoginAct::class.java))
        this.finish()
      }
    }
    return super.onOptionsItemSelected(item)
  }

  private fun openSettingAct(): Boolean {
    startActivity(Intent(this, SettingsActivity::class.java))
    return true
  }

  private fun getUserLogined(): Boolean {
    return loginRepository!!.isLogin(this)
  }


  override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
    when (requestCode) {
      REQ_LOGIN -> if (resultCode == Activity.RESULT_CANCELED) {
        toast("Login canceled")
      } else {
        loadUserData()
      }
      else -> super.onActivityResult(requestCode, resultCode, data)
    }
  }

  private fun loadUserData() {
    //load user name and server photo's in background....
  }

  override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {

  }

  override fun onPageSelected(position: Int) {
    invalidateOptionsMenu()
    Timber.d("onPageSelected pos $position")
//        val fragment = homePageAdapter?.getItem(position)

  }

  override fun onResume() {
    super.onResume()


  }

  override fun onPause() {
    super.onPause()
  }

  override fun onDestroy() {
    super.onDestroy()
  }


  override fun onPageScrollStateChanged(state: Int) {

  }


  companion object {
    private val REQ_LOGIN = 111
    private val REQ_LOCATION_UPDATE = 3333

  }
}

